dlsonic/dlsonic.C

```diff
@@ -0,0 +1,696 @@
+    int dlvsymWithStaticString = 0;
+    int dlmopenWithStaticString = 0;
+
+    // these are to count the # dlsym/dlvsym calls that can be matched
+    // with a corresponding dlopen call.
+    int dlsymMapped = 0;
+    int dlvsymMapped = 0;
+
+    // ideally dlsymWithConstHandl == dlsymWithRTLD_NEXT + dlsymWithRTLD_DEFAULT
+    int dlsymWithConstHandle = 0;
+    int dlsymWithRTLD_NEXT = 0;
+    int dlsymWithRTLD_DEFAULT = 0;
+    
+    static Stats& Instance() {
+        static Stats obj;
+        return obj;
+    }
+
+    void print()
+    {
+#define STAT_FIX_STR(x) (#x "=" + std::to_string(x)) 
+        std::cout << "DIGEST:" << filename << "=["
+                  << STAT_FIX_STR(dlopenCount) << "|"
+                  << STAT_FIX_STR(dlopenWithStaticString) << "|"
+                  << STAT_FIX_STR(dlsymCount) << "|"
+                  << STAT_FIX_STR(dlsymWithStaticString) << "|"
+                  << STAT_FIX_STR(dlvsymCount) << "|"
+                  << STAT_FIX_STR(dlvsymWithStaticString) << "|"
+                  << STAT_FIX_STR(dlmopenCount) << "|"
+                  << STAT_FIX_STR(dlmopenWithStaticString) << "|"
+                  << STAT_FIX_STR(dlsymMapped) << "|"
+                  << STAT_FIX_STR(dlsymWithConstHandle) << "|"
+                  << STAT_FIX_STR(dlsymWithRTLD_NEXT) << "|"
+                  << STAT_FIX_STR(dlsymWithRTLD_DEFAULT) << "|"
+                  << STAT_FIX_STR(dlvsymMapped)
+                  << "]\n";
+#undef STAT_FIX_STR
+    }
+    
+private:
+    Stats() {}
+};
+
+// Store global state helpful in processing
+struct GlobalData
+{
+    // .plt and .plt.sec section limits
+    int64_t pltStartAddr = 0;
+    int64_t pltEndAddr = 0;
+    int64_t pltSecStartAddr = 0;
+    int64_t pltSecEndAddr = 0;
+    // haven't seen a pltGot example but anyway:
+    int64_t pltGotStartAddr = 0;
+    int64_t pltGotEndAddr = 0;
+
+    // index to identify particular calls to dlopen and dlsym
+    uint32_t index = 0;
+
+    enum class CallType {
+        DLOPEN,
+        DLSYM,
+        DLMOPEN,
+        DLVSYM
+    };
+
+    enum class DlHandleType {
+        CONST_RTLDDEFAULT,
+        CONST_RTLDNEXT,
+        CONST_UNKNOWN,
+        CUSTOM // this means handle comes from dlopen
+    };
+
+    // We capture this information each time we find a call to one of the functions
+    // we are tracking.
+    struct CallDetail {
+        uint64_t id = 0;
+        CallType ctype;
+        DlHandleType htype = DlHandleType::CUSTOM;
+        std::optional<uint32_t> mappedTo;
+        uint64_t addr = 0;
+        std::vector<std::string> paramStrs;
+    };
+
+    std::vector<CallDetail> calldetails;
+
+    // used to map dlsym calls to corresponding dlopen calls
+    std::map<uint32_t, std::vector<Dyninst::Node::Ptr>> dlsymIndex2RDISlice;
+    std::map<uint32_t, std::vector<std::pair<uint32_t, uint32_t>>> dlopenIndex2CallFTBlock;
+
+    // used while traversing the block graph
+    std::map<uint32_t, bool> seen;
```

Comment:
uint32_t -> Address

Address is the dyninst type for an address and is an alias for uint64_t. You should use this so the code properly handles >4GiB object files